From 0d743a7d946fe176a07baf2586a6af0e867fd89c Mon Sep 17 00:00:00 2001
From: H.J. Lu <hongjiu.lu@intel.com>
Date: Wed, 11 May 2011 16:19:55 +0200
Subject: [PATCH] Recompute CPU usage at microsecond level

If job finishes quickly, miliseconds arithmetic rounds to zero. If
that happens, recalculate CPU usage ratio with microsecond accuracy
to raise chance to get non-zero values.
---
 resuse.h |    2 ++
 time.c   |   15 +++++++++++++++
 2 files changed, 17 insertions(+), 0 deletions(-)

diff -Naur time-1.7/resuse.h time-1.7.tpg/resuse.h
--- time-1.7/resuse.h	2016-02-06 12:51:23.000000000 +0000
+++ time-1.7.tpg/resuse.h	2016-02-06 12:54:48.694781831 +0000
@@ -33,9 +33,11 @@
 #if HAVE_SYS_RUSAGE_H
 /* This rusage structure measures nanoseconds instead of microseconds.  */
 # define TV_MSEC tv_nsec / 1000000
+# define TV_USEC tv_nsec / 1000
 # include <sys/rusage.h>
 #else
 # define TV_MSEC tv_usec / 1000
+# define TV_USEC tv_usec
 # if HAVE_SYS_RESOURCE_H
 #  include <sys/resource.h>
 # endif
diff -Naur time-1.7/time.c time-1.7.tpg/time.c
--- time-1.7/time.c	2016-02-06 12:51:23.000000000 +0000
+++ time-1.7.tpg/time.c	2016-02-06 12:56:20.822247795 +0000
@@ -331,6 +331,8 @@
 {
   unsigned long r;		/* Elapsed real milliseconds.  */
   unsigned long v;		/* Elapsed virtual (CPU) milliseconds.  */
+  unsigned long ru;            /* Elapsed real microseconds.  */
+  unsigned long vu;            /* Elapsed virtual (CPU) microseconds.  *
 
   if (WIFSTOPPED (resp->waitstatus))
     fprintf (fp, "Command stopped by signal %d\n",
@@ -353,6 +355,17 @@
   v = resp->ru.ru_utime.tv_sec * 1000 + resp->ru.ru_utime.TV_MSEC +
     resp->ru.ru_stime.tv_sec * 1000 + resp->ru.ru_stime.TV_MSEC;
 
+  if (r == 0 && v == 0)
+    {
+      ru = resp->elapsed.tv_usec;
+      vu = resp->ru.ru_utime.TV_USEC + resp->ru.ru_stime.TV_USEC;
+    }
+  else
+    {
+      ru = 0;
+      vu = 0;
+    }
+
   while (*fmt)
     {
       switch (*fmt)
@@ -411,6 +424,8 @@
 	      /* % cpu is (total cpu time)/(elapsed time).  */
 	      if (r > 0)
 		fprintf (fp, "%lu%%", (v * 100 / r));
+             else if (ru > 0)
+               fprintf (fp, "%lu%%", (vu * 100 / ru))
 	      else
 		fprintf (fp, "?%%");
 	      break;

--
1.7.4.4
